pkg/util/hostname/providers.go

```diff
@@ -3,47 +3,193 @@
+
+type provider struct {
+	name string
+	cb   providerCb
+
+	// Should we stop going down the list of provider if this one is successful
+	stopIfSucessful bool
+
+	// expvarName is the name to use to store the error in expvar. This is legacy behavior to match the expvar name
+	// from the previous hostname detection logic.
+	expvarName string
+}
 
 // providerCatalog holds all the various kinds of hostname providers
-var providerCatalog = make(map[string]Provider)
+//
+// The order if this list matters:
+// * Config (`hostname')
+// * Config (`hostname_file')
+// * Fargate
+// * GCE
+// * Azure
+// * container (kube_apiserver, Docker, kubelet)
+// * FQDN
+// * OS hostname
+// * EC2
+var providerCatalog = []provider{
+	{
+		name:            configProvider,
+		cb:              fromConfig,
+		stopIfSucessful: true,
+		expvarName:      "'hostname' configuration/environment",
+	},
+	{
+		name:            "hostname_file",
+		cb:              fromHostnameFile,
+		stopIfSucessful: true,
+		expvarName:      "'hostname_file' configuration/environment",
+	},
+	{
+		name:            "fargate",
+		cb:              fromFarget,
+		stopIfSucessful: true,
+		expvarName:      "fargate",
+	},
+	{
+		name:            "gce",
+		cb:              fromGCE,
+		stopIfSucessful: true,
+		expvarName:      "gce",
+	},
+	{
+		name:            "azure",
+		cb:              fromAzure,
+		stopIfSucessful: true,
+		expvarName:      "azure",
+	},
 
-// RegisterHostnameProvider registers a hostname provider as part of the catalog
-func RegisterHostnameProvider(name string, p Provider) {
-	providerCatalog[name] = p
+	// The following providers are coupled. Their behavior change following the result of the previous provider.
+	// Therefore 'stopIfSucessful' is set to false.
+	{
+		name:            "fqdn",
+		cb:              fromFQDN,
+		stopIfSucessful: false,
+		expvarName:      "fqdn",
+	},
+	{
+		name:            "container",
+		cb:              fromContainer,
+		stopIfSucessful: false,
+		expvarName:      "container",
+	},
+	{
+		name:            "os",
+		cb:              fromOS,
+		stopIfSucessful: false,
+		expvarName:      "os",
+	},
+	{
+		name:            "aws", // ie EC2
+		cb:              fromEC2,
+		stopIfSucessful: false,
+		expvarName:      "aws",
+	},
 }
 
-// GetProvider returns a Provider if it was register before.
-func GetProvider(providerName string) Provider {
-	if provider, found := providerCatalog[providerName]; found {
-		return provider
+// IsConfigurationProvider returns true if the hostname was found through the configuration file
+func (h Data) IsConfigurationProvider() bool {
+	return h.Hostname == configProvider
+}
+
+func saveHostname(cacheHostnameKey string, hostname string, providerName string) Data {
+	data := Data{
+		Hostname: hostname,
+		Provider: providerName,
+	}
+
+	cache.Cache.Set(cacheHostnameKey, data, cache.NoExpiration)
+	if providerName != "" && providerName != "fargate" {
```

Comment:
Why is "fargate" singled out here? (maybe just add a comment)

Reply:
It's like this on main (https://github.com/DataDog/datadog-agent/blob/main/pkg/util/hostname.go#L205). I think there is no concept of host with fargate.

I'll add a comment